src/main/java/org/mule/extension/sftp/internal/proxy/ProtocolState.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.extension.sftp.internal.proxy;
+
+import org.apache.sshd.common.io.IoSession;
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.common.util.buffer.BufferUtils;
+import org.mule.extension.sftp.internal.proxy.socks5.Socks5ClientConnector;
+import org.mule.extension.sftp.internal.proxy.socks5.SocksAuthenticationMethod;
+
+import java.io.IOException;
+
+import static java.text.MessageFormat.format;
+
+public enum ProtocolState {
+
+  NONE,
+
+  INIT {
+
+    @Override
+    public void handleMessage(Socks5ClientConnector connector,
+                              IoSession session, Buffer data)
+        throws Exception {
+      connector.versionCheck(data.getByte());
+      SocksAuthenticationMethod authMethod = connector.getAuthMethod(
+                                                                     data.getByte());
+      switch (authMethod) {
+        case ANONYMOUS:
+          connector.sendConnectInfo(session);
+          break;
+        case PASSWORD:
+          connector.doPasswordAuth(session);
+          break;
+        case GSSAPI:
+          connector.doGssApiAuth(session);
+          break;
+        default:
+          throw new IOException(format("Cannot authenticate to proxy", connector.proxyAddress));
+      }
```

Comment:
This segment is completely procedural. An Object Oriented approach should delegate to the classes. In this case you get the authMethod from the connector and based on the value of that property, you end up invoking another method in the same object. It seems that the decision should be taken inside the `connector` instance.

Consider refactoring to have strategies instead of methods, and have the connector call a method `authenticate(session)` or something similar, and have the right strategy invoked depending of its internal state, rather than extracting the state, and making the decision externally.